pkg/scheduler/queue/tree_queueing_algorithms.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+// QueuingAlgorithm represents the set of operations specific to different approaches to queuing/dequeuing. It is
+// applied at the layer-level -- every Node at the same depth in a MultiQueuingAlgorithmTreeQueue shares the same QueuingAlgorithm,
+// including any state in structs that implement QueuingAlgorithm.
+type QueuingAlgorithm interface {
+	// addChildNode creates a child Node of parent. QueuePaths passed to enqueue functions are allowed to contain
+	// node names for nodes which do not yet exist; in those cases, we create the missing nodes. Implementers are
+	// responsible for adding the newly-created node to whatever state(s) keep track of nodes in the subtree, as well
+	// as updating any positional pointers that have been shifted by the addition of a new node.
+	addChildNode(parent, child *Node)
+
+	// dequeueSelectNode, given a node, returns the node which an element will be dequeued from next, as well
+	// as a bool representing whether the passed node and all its children have been checked for dequeue-ability.
+	// The returned node may be the same as the node passed in; this triggers the base case for the recursive
+	// Node.dequeue, and results in dequeuing from the node's local queue. dequeueSelectNode does *not*
+	// update any Node fields.
+	dequeueSelectNode(node *Node) (*Node, bool)
+
+	// dequeueUpdateState is called after we have finished dequeuing from a node. When a child is left empty after the
```

Comment:
> dequeueUpdateState performs cleanup

maybe
> dequeueUpdateState implementations (may|should) perform cleanup by deleting the empty child Node from the tree and updating internal state to reflect the child Node's removal

It's not a requirement at this point, sort of just an optimization that we think is better than the other option.
Option 1: Deleting empty nodes
* less memory for empty nodes
* depending on queue algo, fewer CPU cycles checking empty nodes for dequeue-ability

Option 2: Not deleting empty nodes
* more CPU cycles deleting nodes and then potentially creating them again when they return

Reply:
also needs something about resetting internal checkedAllNodes-type state after a successful dequeue

Reply:
> also needs something about resetting internal checkedAllNodes-type state after a successful dequeue

Just FYI, this is no longer handled by `dequeueUpdateState` (from your earlier feedback).